site/_posts/2020-05-27-CSI-integration.md

```diff
@@ -0,0 +1,201 @@
+
+Install Velero by running the below command
+
+```bash
+velero install \ 
+--provider azure \ 
+--plugins velero/velero-plugin-for-microsoft-azure:v1.1.0,velero/velero-plugin-for-csi:v0.1.1 \ 
+--bucket $BLOB_CONTAINER \ 
+--secret-file <PATH_TO_CREDS_FILE>/aks-creds \ 
+--backup-location-config resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,storageAccount=$AZURE_STORAGE_ACCOUNT_ID,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID \ 
+--snapshot-location-config apiTimeout=5m,resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID \ 
+--image velero/velero:v1.4.0 \ 
+--features=EnableCSI 
+```
+
+NOTE: Please use the latest released version of the container images from the [official Velero dockerhub](https://hub.docker.com/u/velero).
+
+## Deploy a CSI application
+
+Before installing the CSI application, install the storage class and the volume snapshot class for the Azure disk CSI driver by applying the below `yaml` to our cluster.
+
+```yaml
+apiVersion: storage.k8s.io/v1 
+kind: StorageClass 
+metadata: 
+  name: disk.csi.azure.com 
+provisioner: disk.csi.azure.com 
+parameters: 
+  skuname: StandardSSD_LRS 
+reclaimPolicy: Delete 
+volumeBindingMode: Immediate 
+allowVolumeExpansion: true
+---
+apiVersion: snapshot.storage.k8s.io/v1beta1 
+kind: VolumeSnapshotClass 
+metadata: 
+  name: csi-azuredisk-vsc 
+driver: disk.csi.azure.com 
+deletionPolicy: Retain 
+parameters:
+  tags: 'foo=aaa,bar=bbb' 
+```
+
+NOTE: The above `yaml` was sourced from [StorageClass](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/example/storageclass-azuredisk-csi.yaml) and [VolumeSnapshotClass](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/deploy/example/snapshot/storageclass-azuredisk-snapshot.yaml).
+
+
+Deploy a CSI application: a pod using a CSI backed PVC, in the `csi-app` namespace.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  creationTimestamp: null
+  name: csi-app
+---
+kind: Pod
+apiVersion: v1
+metadata:
+  namespace: csi-app
+  name: csi-nginx
+spec:
+  nodeSelector:
+    kubernetes.io/os: linux
+  containers:
+    - image: nginx
+      name: nginx
+      command: [ "sleep", "1000000" ]
+      volumeMounts:
+        - name: azuredisk01
+          mountPath: "/mnt/azuredisk"
+  volumes:
+    - name: azuredisk01
+      persistentVolumeClaim:
+        claimName: pvc-azuredisk
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  namespace: csi-app
+  name: pvc-azuredisk
+spec:
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+  storageClassName: disk.csi.azure.com
+---
+```
+
+`kubectl apply 👆 yaml`
+
+Once our pod is running write some data into `/mnt/azuredisk`, where the CSI volume is mounted.
```

Comment:
You may want to explain why you're writing data into the file instead of relying on an application's data. Is it ultimately so that you can get a consistent checksum value for purposes on demonstration?

Reply:
yes exactly. I want to show that the data we restored is exactly the same data we backed up.